scheduler/src/cook/kubernetes/compute_cluster.clj

```diff
@@ -163,192 +42,120 @@
+              (count all-task-id->task) "tasks in expected state.")
+    (doseq [[k v] all-task-id->task]
+      (log/debug "Setting expected state for " k " ---> " (task-ent->expected-state v)))
+    (into {}
+          (map (fn [[k v]] [k (task-ent->expected-state v)]) all-task-id->task))))
+
+(defrecord KubernetesComputeCluster [^ApiClient api-client name entity-id match-trigger-chan exit-code-syncer-state
+                                     current-pods-atom current-nodes-atom expected-state-map existing-state-map
+                                     pool->fenzo-atom]
   cc/ComputeCluster
   (launch-tasks [this offers task-metadata-seq]
-    (let [api (CoreV1Api. api-client)
-          pods (map task-metadata->pod task-metadata-seq)]
-      (doseq [pod pods]
-        (log/debug "Launching pod" pod)
-        (try
-          (.createNamespacedPod api "cook" pod nil nil nil)
-          (catch ApiException e
-            (log/error e "Error creating pod:" (.getResponseBody e))
-            (throw e))))))
+    (let [api (CoreV1Api. api-client)]
+      (doseq [task-metadata task-metadata-seq]
+        (controller/update-expected-state
+          this
+          (:task-id task-metadata)
+          {:expected-state :expected/starting :launch-pod (api/task-metadata->pod task-metadata)}))))
 
   (kill-task [this task-id]
-    (throw (UnsupportedOperationException. "Cannot kill tasks")))
+    (controller/update-expected-state this task-id {:expected-state :expected/killed}))
 
-  (decline-offer [this offer-id])
+  (decline-offers [this offer-ids]
+    (log/debug "Rejecting offer ids" offer-ids))
 
   (db-id [this]
     entity-id)
 
   (compute-cluster-name [this]
     name)
 
-  (initialize-cluster [this pool->fenzo pool->offers-chan]
+  (initialize-cluster [this pool->fenzo running-task-ents]
+    ; Initialize the pod watch path.
     (let [conn cook.datomic/conn
-          handle-exit-code (fn handle-exit-code [task-id exit-code]
-                             (sandbox/aggregate-exit-code exit-code-syncer-state task-id exit-code))
-          handle-status-update (fn handle-status-update [status]
-                                 (scheduler/handle-status-update conn
-                                                                 this
-                                                                 pool->fenzo
-                                                                 (constantly nil) ; no sandboxes to sync
-                                                                 status))
-          pod-callback (make-pod-watch-callback handle-status-update handle-exit-code)]
-      (initialize-pod-watch api-client current-pods-atom pod-callback))
-    (initialize-node-watch api-client current-nodes-atom)
+          pod-callback (make-pod-watch-callback this)]
+      (api/initialize-pod-watch api-client current-pods-atom pod-callback)
+      ; We require that initialize-pod-watch sets current-pods-atom before completing.
```

Comment:
Fix comment. ""Before executing the next line.""

Reply:
I'd vote for "before calling determine-expected-state" as opposed to "next line"

Reply:
Actually, what about just killing the comment and adding a check in `determine-expected-state`?

Reply:
Because the state could legit be empty.